lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,410 @@
+        ));
+
+        _cancelTransitionEventListener();
+
+        complete();
+      });
+
+      _endTransitionSubscription.current =
+          transitionNodeRef.current?.onTransitionEnd?.skip(transitionCount - 1)?.take(1)?.listen((_) {
+        _cancelTransitionEndTimer();
+
+        complete();
+      });
+    }
+
+    // ----- Effects (Lifecycle) ----- //
+
+    final _isMounted = useRef(false);
+
+    // Mount + Unmount
+    useEffect(() {
+      // componentDidMount
+      _isMounted.current = true;
+
+      // componentWillUnmount
+      return () {
+        _isMounted.current = false;
+
+        _cancelTransitionEndTimer();
+        _cancelTransitionEventListener();
+      };
+    }, const []);
+
+    // componentDidUpdate
+    useLayoutEffect(() {
+      if (!_isMounted.current) return;
+      _transitionNotGuaranteed.current = false;
+
+      if (transitionPhase.value != TransitionPhase.SHOWING) {
+        // Allows the AbstractWithTransitionComponent to handle state changes that interrupt state
+        // changes waiting on transitionend events.
+        _cancelTransitionEventListener();
+      }
+
+      /// Called when `transitionPhase` is `TransitionPhase.PRE_SHOWING`.
+      void _handlePreShowing() {
+        onNextTransitionEnd(() {
+          if (_isOrWillBeShown(transitionPhase.value)) {
+            transitionPhase.set(TransitionPhase.SHOWN);
+          }
+        });
+
+        transitionPhase.set(TransitionPhase.SHOWING);
+      }
+
+      /// Called when `transitionPhase` is `TransitionPhase.HIDING`.
+      void _handleHiding() {
+        if (_transitionNotGuaranteed.current) {
+          // No transition will occur, so kick off the state change manually.
+          transitionPhase.set(TransitionPhase.HIDDEN);
+
+          _cancelTransitionEventListener();
+          _cancelTransitionEndTimer();
+        } else {
+          onNextTransitionEnd(() {
+            if (_isMounted.current && transitionPhase.value == TransitionPhase.HIDING) {
+              transitionPhase.set(TransitionPhase.HIDDEN);
+            }
+          });
+        }
+      }
+
+      switch (transitionPhase.value) {
+        case TransitionPhase.PRE_SHOWING:
+          _handlePreShowing();
+          break;
+        case TransitionPhase.SHOWING:
+          // NOTE: props.onWillShow is called within _handleShow in the handler that
+          // triggers an update to `transitionPhase` that will result in this effect being processed.
+          break;
+        case TransitionPhase.HIDING:
+          // NOTE: props.onWillHide is called within _handleHide in the handler that
+          // triggers an update to `transitionPhase` that will result in this effect being processed.
+          _transitionNotGuaranteed.current = transitionPhase.value == TransitionPhase.SHOWING;
+          _handleHiding();
+          break;
+        case TransitionPhase.HIDDEN:
+          props.onDidHide?.call();
```

Comment:
I believe these need to also need to be included in the dependencies array: https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect

Reply:
Addressed in https://github.com/Workiva/over_react/pull/621/commits/967d93dd4c6cbb6fac367b02d128a6002613d8b8